### PR TITLE
refactor: reorganize TS compiler

### LINKED ISSUE #5507: Typescript compile error during bundle gives unhelpful assertion error
In this case, I forgot to pass `-c tsconfig.json` to add DOM to the libs. I'm assuming this should be instead outputting the compile errors.
```
PS C:\Users\mark\Documents\PROJECTS\deno-test> deno --version
deno 1.0.0
v8 8.4.300
typescript 3.9.2
PS C:\Users\mark\Documents\PROJECTS\deno-test> deno bundle ./src/index.ts
Bundling file:///C:/Users/mark/Documents/PROJECTS/deno-test/src/index.ts
error: Uncaught AssertionError: Assertion failed.
    at Object.assert ($deno$/util.ts:33:11)
    at compile ($deno$/compiler.ts:1355:5)
    at async tsCompilerOnMessage ($deno$/compiler.ts:1548:22)
    at async workerMessageRecvCallback ($deno$/runtime_worker.ts:74:9)
```

### PR DESCRIPTION
**Pull request by bartlomieju**
Yet another refactor of TS compiler:
- made `bundle()` a standalone function instead of method of `TsCompiler`
- TS compiler doesn't depend on TextEncoder/TextDecoder
- removed dead code from TS compiler
- fixed typos

Fixes  #5507

### CODE REVIEW

**Comment by bartlomieju** [comment]
> LGTM so far... let me know if you want/need anything

@kitsonk I want TS worker to not use any console logs - so I want to move `processConfigureResponse` to Rust; can you look into that?

**Comment by kitsonk** [comment]
> `processConfigureResponse`

That is going to be hard-ish because it relies on `ts` APIs to ensure that stuff can be parsed, and it supports JSONC, and serde specifically refuses to support JSONC.  I will have to look at swc and see if they support parsing of `tsconfig.json` files easily enough.  Will look into it.

> EDIT: commonPath and normalizeUrl also look shady... If we can remove them/move logic into Rust

They aren't shady! :-) I am not too sure we want to move them, because they are tightly coupled to how TypeScript generates the bundle (as in you have to "guess" at what the module specifiers are going to be because of this: https://github.com/microsoft/TypeScript/issues/35052).  They should go when we move the bundle generation out to Rust (as in all we do is take the single outfile and do the munging in Rust) which doesn't seem like we are ready to do that quite yet.

**Comment by bartlomieju** [comment]
> > `processConfigureResponse`
> 
> That is going to be hard-ish because it relies on `ts` APIs to ensure that stuff can be parsed, and it supports JSONC, and serde specifically refuses to support JSONC. I will have to look at swc and see if they support parsing of `tsconfig.json` files easily enough. Will look into it.

`dprint` supports parsing of JSONC https://github.com/dprint/dprint/commit/ce9b5a6b3eda2810331b24456e47551f64128e3c - maybe @dsherret can expose parser?

> > EDIT: commonPath and normalizeUrl also look shady... If we can remove them/move logic into Rust
> 
> They aren't shady! :-) I am not too sure we want to move them, because they are tightly coupled to how TypeScript generates the bundle (as in you have to "guess" at what the module specifiers are going to be because of this: [microsoft/TypeScript#35052](https://github.com/microsoft/TypeScript/issues/35052)). 

I see, thanks for explanation I couldn't really grok what they are for.

> They should go when we move the bundle generation out to Rust (as in all we do is take the single outfile and do the munging in Rust) which doesn't seem like we are ready to do that quite yet.

How would it take? I already do some munching on source maps generated during runtime code compilation so that shouldn't be a big deal.

**Comment by dsherret** [comment]
@bartlomieju the JSONC parser is here: https://github.com/dprint/jsonc-parser

I just recently added the `parse_to_value` function that you could use here (right now it wastes time also collecting comments and tokens internally, but I'll improve it to not do that soon).

**Comment by kitsonk** [comment]
I broke out the other two things into #5606 and #5607 so I could explain a bit more what is going on or needs to go on in my opinion as well as make sure it doesn't get lost.  _small bites_

**Comment by bartlomieju** [comment]
@ry let's land this PR for 1.0.1

File: cli/js/compiler.ts [hunk 1]
```diff
@@ -673,7 +632,7 @@ function buildSourceFileCache(
   }
 }
 
-interface EmmitedSource {
+interface EmittedSource {
```
Comment (ry) [inline cli/js/compiler.ts:635]: Yikes.

Comment (bartlomieju) [reply cli/js/compiler.ts:635]: Those are all private interfaces not visible to end users

**Pull request merged**